### ISSUE TITLE
[BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding (2019)]

### ISSUE THREAD

**Issue by sallyy1** [sallyy1/NLU-NLP#16]
## 요약 및 Contribution
- BERT : **Bidirectional Encoder Representations from Transformers.**
- BERT는 모든 mulit-layer에서 left와 right context(문맥)을 모두 활용하는 deep bidirectional 모델
- 대용량의 unlabeled data로 pre-training하고 특정 task에 대해 transfer learning하는 모델
- 이때, task-specific한 구조를 디자인 할 필요가 없음. (Unified architecture across different tasks로 pre-trained 아키텍처와 최종 down-stream Task 아키텍처 간에는 최소한의 차이만 있음)

> As a result, the pre-trained BERT model can be finetuned with just one additional output layer to create state-of-the-art models for a wide range of tasks, such as question answering and language inference, without substantial task-specific architecture modifications.

## Motivation
- unidirectional이 아닌 bidirectional
- MLM과 NSP를 활용한 pretraining으로 token 간 (양방향) 문맥과 sentence간 연관성을 모두 고려할 수 있는 모델
- pre-trained 모델과 같은 파라미터가 다른 down-stream Task에서 모델을 Initalize하는 데 쓰임
- sentence-level, token-level 모두에서 범용적 자연어 task로의 활용 가능

#### 1. two strategies for applying pre-trained language representation to down-stream tasks (Unsupervised)
- 기존 Pre-trained 표현력에 제한이 있었음.
- uni-directinoal이라는 점

##### 1) feature-based approach
- **ELMo** : task-specific된 아키텍처를 사용하며, pre-trained representation을 부가적인 feature로 사용
- 전통적인 단어 임베딩 연구를 다른 차원에서 일반화한 연구 ("Contextual word embeddings")
- 독립적인 'a left-to-right'과 'a right-to-left' 언어모델로 'context-sensitive features'를 추출/생성하여, 이 둘의 concat으로 각 토큰의 문맥을 표현.
- 양방향(->과 <-)이나, shallow한 bi-directional

- **Melamud et al. (2016)** : 독립적인 left-to-right'과 'a right-to-left 컨텍스트를 LSTM로 결합하여 사용. 하나의 단어를 예측하는 작업을 통해 'contextual representations'를 학습

- **BERT** : deep bi-directional. 모든 레이어에서 bi-directional한 방식으로 학습

##### 2) fine-tuning approach
- **OpenAI GPT**(Generative Pre-trained Transformer) : 최소한의 task-specific 파라미터를 모두 fine-tunning 해 down-stream task에 맞게 훈련하여 사용
- 'a left-to-right architecture'
- 한 방향(->), 모든 토큰은 트랜스포머의 self-attention layers에서 이전 토큰의 영향만 받을 수 있다. 이 제한은 sentence-level Task에서만 차선이며, 양방향의 문맥을 통합하는 것이 중요한 token-level Task 에서는 fine-tuning approach를 적용할 때 좋지 않다.
- 제한된 self-attention
- 모든 fine-tuning experiments에 5e-5로 같은 러닝레이트를 줌

- **BERT** : bi-directional self-attention & 범용적인 모델을 만들기 위해, 각 특정 task에 특화된 task-specific한 fine-tunning 러닝레이트를 선택

#### 2. MLM 관련
- **Cloze task (Taylor, 1953)** : MLM을 pre-training에 사용하였으나, uni-directional 한 것에 제한됨

- **BERT** : bi-directional 한 MLM

#### 3. NSP 관련
- NSP : Next Sentence Prediction
- **representationlearning objectives** (Jernite et al. (2017) and Logeswaran and Lee (2018)) : 'sentence embedding만 down-stream tasks transfer 과정에 쓰였음
- **GPT** : [SEP]와 [CLS]를 fine-tuning 시에만 사용
- **BERT** : 모든 파라미터를 최종 task 모델 파라미터에 initialize하여 transfer & [SEP]와 [CLS]를 pre-training의 문장 임베딩 시에도 사용

#### 1) **"sentence-level" Task** : 'natural language inference', 'paraphrasing' 등이 있으며, 목적은 문장을 전체적으로 분석함으로써 문장 간 연관성을 예측하는 것
(RNN과 같은 모델에서 모든 time-step에 대해 입력과 출력이 한번에 입출력 되는 모델을 사용하기에 'sentence-level'이라고 표현하는 것 같다.) 
- 'Non-auto-regressive'한 Task (=Non-generative) : 현재 상태가 앞 뒤 상태를 통해 정해지는 경우 (예: Part of Speech(POS) Tagging, Text Classification)
- Bidirecional RNN 사용이 권장됨

#### 2) **"token-level" Task** : 'named entity recognition', 'question answering' 등, token 레벨에서 세분화된(fine-grained) 출력을 생성해야 하는 모델
(seq2seq(의 Decoder)와 같은 모델에서 각 time-step별로 입력이 들어와서 sequentional하게 출력이 나오는 one-to-many와 같은 모델을 사용하기에 'token-level'이라고 표현하는 것 같다.)
- 'Auto-regressive'한 Task (=Generative) : 현재 상태가 과거 상태에 의존하여 정해지는 경우 (예: Natural Language Generation, Machine Translation)
- 'Conditional language model(조건부)'과 관련
- Bidirectional RNN 사용 불가, Uni-birectional RNN 사용

#### * Masked language model(MLM)
- Input에서 일부 tokens을 무작위로 마스킹하며, 그 목적은 문맥(context) 기반으로 마스킹된 단어의 원래 어휘 id를 예측하는 것

## Model
- BERT의 Framework는 크게 two step : 1) pre-training, 2) fine-tuning
- unlabeled한 sentence A와 B 쌍을 pre-training 해서 -> token or sentence level Task를 bi-directional하게 해결 가능
- 1) pre-training Task는 크게 1) MLM과 2) NSP
- 2) fine-tuning은 Transformer의 self-attention 매커니즘인데, bidirectional cross attention

###### 1) Masked LM
- 효과 : 문맥을 활용할 수 있는 모델 (특히, 양방향 문맥(context)를 통합하는 것이 중요.)
- 'token-level' 학습에 해당됨
- pretrain 시에만 입력에 랜덤하게 Mask 주고 fine-tunning 에서는 사용하지 않음
- 마스킹 룰: 80%는 [MASK] 토큰으로 교체, 10%는 랜덤 단어로 교체, 10%는 그대로 유지하며 예측되어야 하는 실제 단어

###### 2) Next Sentence Prediction(NSP)
- 효과: 두 문장 간의 relationship을 파악하는 모델
- 'sentence-level' 학습에 해당됨
- 데이터셋에서 A, B 두개의 문장을 구성하는데, 50%는 실제 이어지는 문장이며 'IsNext'로 레이블링, 50%는 랜덤한 다음 문장이며, 'NotNext'로 레이블링하여 이진 분류
- 간단한 것에 비해, QA, NLI 모두에서 도움이 됨을 확인함

#### 2nd Step : Fine-tunning Bert
- [CLS] 토큰과 [SEP] 토큰을 줌으로써, token & sentence level Task에서 모두 활용 가능
- 모든 시퀀스의 시작을 [CLS] 스페셜 토큰으로 줌 (Classification Task에서만 사용되면, 아닌 경우에는 무시됨)
- [SEP] 스페셜 토큰을 문장 사이와 Input의 마지막에 줌 => 'token' 임베딩과 'position' 임베딩에 'segment' 임베딩까지 합쳐져  Input representation 이 됨 (각 토큰에 대한 Input Embedding은 세 임베딩의 합)
- ex) Classification Task을 할 때에는 _맨 첫 번째 자리의 transformer의 output을 활용_(many-to-one 측면?)
- ex) QA Task을 할 때에는 첫 번째 문장에 question, 두 번째 문장에 paragraph를 넣어서 입력데이터를 설정 -> QA Task에서 answer문장과 paragraph문장의 사작과 끝 vector를 학습 -> 각 vector는 output과 dot product+softmax 후 cross-entropy를 통해 학습

-----------------------------------------
https://arxiv.org/pdf/1810.04805.pdf&usg=ALkJrhhzxlCL6yTht2BRmH9atgvKFxHsxQ
https://blog.naver.com/PostView.nhn?blogId=winddori2002&logNo=222022178447&categoryNo=32&parentCategoryNo=1&viewDate=&currentPage=1&postListTopCurrentPage=1&from=postView
https://dnddnjs.github.io/nlp/2019/05/08/BERT/
